added source-tar option

pizzapanther opened the pull request:
I added a `source-tar` option so that you could build a custom tar locally without having to upload it somewhere. I wanted to be able to add custom excludes during the tar generation.

If this looks good, I'll add the docs before merge. I wanted to get some feedback first to see if this is a good direction.

Sample deploy script I use now:

```bash
#!/bin/bash

rm deploy.tar.gz
tar cvf deploy.tar *.j* public/ src/
gzip -9 deploy.tar
heroku builds:create -a my-app-name --source-tar deploy.tar.gz
```

dmathieu commented on commands/builds/create.js:
```diff
@@ -43,25 +43,32 @@ function compressSource (context, dir, tempFile) {
 
-async function uploadDirToSource (context, heroku, dir) {
-  let tempFilePath = path.join(os.tmpdir(), uuid.v4() + '.tar.gz')
+async function uploadDirToSource (context, heroku, dir, filePath) {
```
We will always have either a filePath or a dir, never both, which are two quite similar arguments. It looks like we could get rid of one.
Possibly by renaming the argument to `path`, and detecting if it's a directory or not and changing the behavior?

Then, the method would also be renamed to `uploadToSource`, as it doesn't only upload a directory.

pizzapanther commented:
@dmathieu merged the arguments, added the test for the new option, added another test for the `source-url` option since that was missing, cleaned up some repeating test code, and updated the docs

dmathieu commented on commands/builds/create.js:
```diff
@@ -43,25 +43,36 @@ function compressSource (context, dir, tempFile) {
 
-  if (!sourceUrl) {
+  if (sourceTar) {
+    sourceUrl = await uploadDirToSource(context, heroku, path.join(dir, sourceTar))
```
What if we passed an absolute path here?

pizzapanther replied:
makes sense, updating ...

pizzapanther commented:
![giphy-downsized](https://user-images.githubusercontent.com/552478/33607722-997c1fd0-d987-11e7-8c89-1095cf4d72f2.gif)
Squashed !

dmathieu commented:
Thank you Paul!

pizzapanther commented:
i'm trying out the whole monolithic repository approach so this addition is invaluable for me.  Thanks for the review and merge. 👍

pizzapanther commented:
What's the release cycle like? When will this show up in the install when using: `heroku plugins:install heroku-builds`?

dmathieu commented:
I've cut a new release right after merging this PR. So the latest version is on npm and should available with `h plugins:update`.

mmiller42 commented:
I only seem to be able to install 0.0.20, even after a `heroku plugins:update` and `npm cache clear --force`.

If I try a fresh install via `heroku plugins:install heroku-builds`, I get 0.0.20 installed, and if I try `heroku plugins:install heroku-builds@0.0.21`, I get the error:

```
error Couldn't find any versions for "heroku-builds" that matches "0.0.21"
```

As far as my understanding is with Heroku, it just pulls the package from the NPM repository, right? I'm wondering if there's just something screwed up on my end, but I can't figure out what it could be.

dmathieu commented:
Ping @heroku/cli

jdxcode commented:
I think there is some kind of bug with npm and/or yarn as this happened in another plugin. Publishing again seems to fix it though.

dmathieu commented:
I've just published v0.0.22. Could you try updating?

jdxcode commented:
@dmathieu it's still not working, let me look into this a bit
